Attn : Dylan :: Malformed Cookie Bug : 17 December 2015

swashcap opened the pull request:
Asana task: https://app.asana.com/0/37131652076106/74571845162511

swashcap committed:
Add 'Auth' service unit tests.

swashcap committed:
Change `Auth.logout()` test to use a callback.

swashcap committed:
Fix log out bug.

This addresses the log out cookie malforming bug by:

* Sending the `authCookieName` option to the `Auth` service for
  later use
* Using the `js-cookie` NPM module for clearer cookie removal, which
  also ensures the cookies don't become malformed
* Removing malformed cookies which cause the API server to send an
  'Invalid Cookie Header' response with every request

swashcap committed:
Fix module name.

cdaringe commented on test/auth.js:
```diff
@@ -0,0 +1,79 @@
+'use strict';
+
+var Auth = require('../scripts/lib/auth');
+var Cookies = require('js-cookie');
```
are `Auth` and `Cookies` constructors?

swashcap replied:
js-cookie uses [`Cookies` in their docs](https://github.com/js-cookie/js-cookie#basic-usage). Not sure why I capitalized `Auth`. I can change it.

swashcap replied:
Fixed in 7d22647c4e20ce8eb2e9a375ff1cb37768c32319.

swashcap committed:
Downcase non-constructor module names.

dylancwood commented on scripts/lib/auth.js:
```diff
@@ -64,23 +80,36 @@ function getApiUrl(endpoint) {
+        path: '/'
+    };
+
+    if (hostPieces.length > 2) {
```
Isn't the domain always the last two `hostPieces`, regardless of length?

swashcap replied:
When testing it’s `localhost`.

dylancwood replied:
doh! Good point. Sorry.

dylancwood commented on scripts/lib/auth.js:
```diff
@@ -64,23 +80,36 @@ function getApiUrl(endpoint) {
+    };
+
+    if (hostPieces.length > 2) {
+        options.domain = '.' + hostPieces.slice(hostPieces.length - 2).join('.');
```
You can just do hostPieces.slice(-2)` to grab the last two elements of an array.

proof:

```
> ['a', 'b', 'c'].slice(-2)
< ["b", "c"]
```

swashcap replied:
Ah, good call. Didn’t know about that.

swashcap replied:
Fixed in 9104157fe2022a8bc8dee78f1516216c2934418c.

dylancwood commented on scripts/lib/auth.js:
```diff
@@ -167,9 +196,11 @@ function login(username, password) {
  * @return {Promise}
  */
-function logout() {
+function logout(cb) {
```
I see that the callback is used in tests, but why can't the test use promises instead of a callback?

dylancwood replied:
What if we just return `jQuery.ajax`?

swashcap replied:
462f9d7d8fa9af60c49d7e9a633327e32d4ef2a8 addresses this.

swashcap committed:
Shorten `slice` arguments.

swashcap committed:
Rewrite auth module's `logout` method.

This makes the `logout` return a jQuery deferred instead of both a
deferred.promise() and callback.
